### PR TITLE
Add XLA to transformer

### PR DESCRIPTION
**Pull request by tfboyd**
- Moved config logic to keras_utils
- Added enable_xla flag to _performance flags
- refactored flags for keras resnet
- created common set_session_config

### CODE REVIEW

File: official/resnet/keras/keras_common.py [hunk 1]
```diff
@@ -419,8 +375,10 @@ def data_delay_prefetch():
 
 
 def set_cudnn_batchnorm_mode():
-  """Set CuDNN batchnorm mode for better performance. Note that the spatial
-     persistent mode may lead to accuracy losses for certain models."""
+  """Set CuDNN batchnorm mode for better performance. Note that the spatial.
```
Comment (reedwm) [inline official/resnet/keras/keras_common.py]: Don't split the sentence starting with "Note that the spatial". Instead, format like:

```
"""Set CuDNN batchnorm mode for better performance.

Note that the spatial persistent mode may lead to accuracy losses for certain
models.
"""

File: official/transformer/transformer_estimator_benchmark.py [hunk 2]
```diff
@@ -100,10 +101,12 @@ def _report_benchmark(self,
       exp_per_sec = sum(exp_per_second_list) / (len(exp_per_second_list))
       metrics.append({'name': 'exp_per_second',
                       'value': exp_per_sec})
-    self.report_benchmark(
-        iters=eval_results['global_step'],
-        wall_time=wall_time_sec,
-        metrics=metrics)
+
+    flags_str = flags_core.get_nondefault_flags_as_str()
```
Comment (reedwm) [inline official/transformer/transformer_estimator_benchmark.py]: Thanks for adding FLAGs to the estimator version!

Now the only thing left (that I really want) is the Resnet keras version.

File: official/utils/flags/_performance.py [hunk 3]
```diff
@@ -270,4 +271,9 @@ def _check_fp16_implementation(flags_dict):
             "Whether to enable tf.data's `experimental_slack` option.")
     )
 
+  if enable_xla:
+    flags.DEFINE_boolean(
```
Comment (reedwm) [inline official/utils/flags/_performance.py]: Maybe also have the Keras Resnet model use this version of the flag, so we don't have two definitions of the `enable_xla` flag.

File: official/transformer/v2/transformer_main.py [hunk 4]
```diff
@@ -121,6 +122,12 @@ def __init__(self, flags_obj):
   def train(self):
     """Trains the model."""
     params, flags_obj, is_train = self.params, self.flags_obj, True
+    # Sets config options.
+    keras_utils.set_config_v2(
```
Comment (reedwm) [inline official/transformer/v2/transformer_main.py]: I think the transformer model supports TF 1. A simple fix would be that if TF1 is used, and either enable_xla==False or enable_grappler_layout_optimizer==True, then raise an error saying those flags are only supported with TF 2.

File: official/utils/flags/_performance.py (hunk 3 above)
Comment (tfboyd) [reply official/utils/flags/_performance.py]: I'll follow your "hack" (pattern) and add it in.  I really want to fix the flags correctly for keras.  I do not like that we call into run_loops to make the flags.

File: official/transformer/v2/transformer_main.py (hunk 4 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: I will add in the TF 1.0 config code.  I was not sure this works with TF 1.0.  I might have been thinking of the estimator version not being upgraded for TF 2.0.

File: official/resnet/keras/keras_common.py (hunk 1 above)
Comment (tfboyd) [reply official/resnet/keras/keras_common.py]: Totally not thinking on my part.

File: official/utils/flags/_performance.py (hunk 3 above)
Comment (reedwm) [reply official/utils/flags/_performance.py]: Yeah, the pattern is pretty ugly. It's why I used the word "Maybe" in this comment :).

File: official/resnet/keras/keras_common.py (hunk 1 above)
Comment (tfboyd) [reply official/resnet/keras/keras_common.py]: fixed

File: official/transformer/v2/transformer_main.py (hunk 4 above)
Comment (tfboyd) [reply official/transformer/v2/transformer_main.py]: I refactored this entirely.  I believe I have thought it through and saved a lot of lines of code and made this easier to reuse.

File: official/utils/flags/_performance.py (hunk 3 above)
Comment (tfboyd) [reply official/utils/flags/_performance.py]: Fixed, I think it is slightly better and ready for a bigger change soon.

File: official/resnet/keras/keras_cifar_main.py [hunk 5]
```diff
@@ -98,17 +99,7 @@ def run(flags_obj):
...
-  if keras_common.is_v2_0():
-    keras_common.set_config_v2()
-  else:
-    config = keras_common.get_config_proto_v1()
-    if flags_obj.enable_eager:
-      tf.compat.v1.enable_eager_execution(config=config)
-    else:
-      sess = tf.Session(config=config)
-      tf.keras.backend.set_session(sess)
+  keras_utils.set_session_config(enable_eager=flags_obj.enable_eager)
```
Comment (reedwm) [inline official/resnet/keras/keras_cifar_main.py]: Why not pass `enable_xla` here?

File: official/utils/misc/keras_utils.py [hunk 6]
```diff
@@ -128,3 +129,73 @@ def on_batch_end(self, batch, logs=None):
...
+        enable_xla=enable_xla,
+        enable_grappler_layout_optimizer=enable_grappler_layout_optimizer)
+    if enable_eager:
+      tf.compat.v1.enable_eager_execution(config=config)
+    else:
+      sess = tf.Session(config=config)
+      tf.keras.backend.set_session(sess)
+
+
+def get_config_proto_v1(enable_xla=False,
```
Comment (reedwm) [inline official/utils/misc/keras_utils.py]: For this function and below, don't give default values to arguments.

File: official/transformer/v2/misc.py [hunk 7]
```diff
@@ -69,13 +69,26 @@ def define_transformer_flags():
...
+           'convolutions and batch normalizations, and this flag allows to '
+           'disable it.'
+  )
+
   flags_core.define_benchmark()
   flags_core.define_device(tpu=True)
 
   flags.DEFINE_integer(
-      name='train_steps', short_name='ts', default=None,
+      name='train_steps', short_name='ts', default=300000,
```
Comment (reedwm) [inline official/transformer/v2/misc.py]: Why change this default value?

File: official/transformer/v2/misc.py [hunk 8]
```diff
@@ -69,13 +69,26 @@ def define_transformer_flags():
...
+      enable_xla=True
   )
+
+  # Additional performance flags
+  # TODO(b/76028325): Remove when generic layout optimizer is ready.
+  flags.DEFINE_boolean(
+      name='enable_grappler_layout_optimizer',
+      default=True,
+      help='Enable Grappler layout optimizer. Currently Grappler can '
+           'de-optimize fp16 graphs byt forcing NCHW layout for all '
```
Comment (reedwm) [inline official/transformer/v2/misc.py]:
```suggestion
           'de-optimize fp16 graphs by forcing NCHW layout for all '
```

File: official/transformer/v2/misc.py (hunk 7 above)
Comment (tfboyd) [reply official/transformer/v2/misc.py]: This was already submitted.  None results in an error.

File: official/utils/misc/keras_utils.py (hunk 6 above)
Comment (tfboyd) [reply official/utils/misc/keras_utils.py]: based on our conversation I think this is the best I can do and I don't want to go back to FLAGs.

File: official/resnet/keras/keras_cifar_main.py (hunk 5 above)
Comment (tfboyd) [reply official/resnet/keras/keras_cifar_main.py]: That flag is not turned on for ResNet56 yet.  I hope to add FP16 and XLA in a future change.
